stdlib/seq.ml

```diff
@@ -62,6 +62,16 @@ let rec flat_map f seq () = match seq () with
 
 let concat_map = flat_map
 
+let rec intersperse sep seq () = match seq () with
+  | Nil -> Nil
+  | Cons (x, next) ->
+      match next () with
+      | Nil -> return x ()
+      | Cons (y, next) -> Cons (x, cons sep (intersperse sep (cons y next)))
```

Comment:
The call `cons y next` re-conses the entire sequence; I suggest the following variation that avoids this extra allocation:
```suggestion
let rec intersperse sep seq () =
  let rec aux seq () =
    match seq () with
    | Nil -> Nil
    | Cons (x, next) -> Cons (sep, cons x (aux next))
  in
  match seq () with
  | Nil -> Nil
  | Cons (x, next) -> Cons (x, aux next)
```

Reply:
Can't we do something like

```ocaml
| Cons _ as next -> Cons (x, cons sep (intersperse sep (fun () -> next)))
```

which should avoid re-consing?